corehq/apps/user_importer/helpers.py

```diff
@@ -188,35 +190,24 @@ def update_name(self, name):
         self.user.set_full_name(str(name))
         self.logger.add_changes({'first_name': self.user.first_name, 'last_name': self.user.last_name})
 
-    def update_user_data(self, data, uncategorized_data, profile, domain_info):
-        # Add in existing data. Don't use metadata - we don't want to add profile-controlled fields.
-        current_profile_id = self.user.user_data.get(PROFILE_SLUG)
-
-        for key, value in self.user.user_data.items():
-            if key not in data:
-                data[key] = value
-        if profile:
-            profile_obj = domain_info.profiles_by_name[profile]
+    def update_user_data(self, data, uncategorized_data, profile_name, domain_info):
+        from corehq.apps.users.user_data import UserDataError
+        user_data = self.user.get_user_data(self.user_domain)
+        old_profile_id = user_data.profile_id
+        if PROFILE_SLUG in data:
+            raise UserUploadError(f"You cannot set {PROFILE_SLUG} directly")
+        if profile_name:
+            profile_obj = domain_info.profiles_by_name[profile_name]
             data[PROFILE_SLUG] = profile_obj.id
-            for key in profile_obj.fields.keys():
-                self.user.pop_metadata(key)
```

Comment:
Is it intentional to remove these lines?

Reply:
Yep, that was deliberate.  The old version of this added all stored user data to the passed in `data` a few lines above this, then this loop removed profile-controlled fields from `user.user_data` (not from `data`).  It then called `update_metadata` with `data`, so anything that had been removed from `user.user_data` is going to be added back in anyways.  So it's a question of profile conflicting with data versus data conflicting with profile.

Honestly, I had a lot of trouble teasing out how this method works, and it turns out this [essentially doesn't work](https://dimagi-dev.atlassian.net/browse/USH-2762) in its current form end-to-end, so I feel like there's some leeway to fiddle with the logic.  My aim was largely to turn this into a more straightforward update, and use the test cases to convince myself that the logic is sound.